File: frontend/fetch_countries.py
import json
import urllib.request
import ssl
import os

ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

req = urllib.request.Request('https://restcountries.com/v3.1/all')
req.add_header('User-Agent', 'Mozilla/5.0 (Windows NT 10.0; Win64; x64)')
req.add_header('Accept', 'application/json')

with urllib.request.urlopen(req, context=ctx) as response:
    data = json.loads(response.read().decode())

formatted = []
for c in data:
    if not isinstance(c, dict): continue
    if 'idd' not in c or not isinstance(c['idd'], dict) or 'root' not in c['idd']: continue
    
    code = c.get('cca2', '')
    nameFr = c.get('translations', {}).get('fra', {}).get('common', '')
    if not nameFr:
        nameFr = c.get('name', {}).get('common', 'Unknown')
        
    nameEn = c.get('translations', {}).get('eng', {}).get('common', '')
    if not nameEn:
        nameEn = c.get('name', {}).get('common', nameFr)
        
    if nameFr == 'Unknown': continue
    
    root = c['idd']['root']
    suffixes = c['idd'].get('suffixes', [])
    dialCode = root + (suffixes[0] if len(suffixes) == 1 else '')
    
    flag = c.get('flag', '')
    
    # Clean dial codes that look like '+1242' etc while keeping the first suffix if applicable
    
    formatted.append({
        'code': code,
        'nameFr': nameFr,
        'nameEn': nameEn,
        'dialCode': dialCode,
        'flag': flag
    })

formatted.sort(key=lambda x: x['nameFr'])

content = """export type CountryData = {
    code: string;
    nameFr: string;
    nameEn: string;
    dialCode: string;
    flag: string;
};

export const countries: CountryData[] = """ + json.dumps(formatted, ensure_ascii=False, indent=4) + ";\n"

out_path = 'src/utils/countries.ts'
os.makedirs(os.path.dirname(out_path), exist_ok=True)
with open(out_path, 'w', encoding='utf-8') as f:
    f.write(content)

print(f"Successfully wrote {len(formatted)} countries to {out_path}")

File: frontend/fetch_countries.js
const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');

function generate() {
    console.log('Fetching stable countries github source...');
    // Increase maxBuffer because countries.json is > 1MB
    const rawData = execSync('curl -s https://raw.githubusercontent.com/mledoze/countries/master/countries.json', { maxBuffer: 1024 * 1024 * 10 }).toString();
    const data = JSON.parse(rawData);

    const formatted = [];
    for (const c of data) {
        if (!c.idd || !c.idd.root) continue;
        
        const code = c.cca2 || '';
        let nameFr = (c.translations && c.translations.fra && c.translations.fra.common) || '';
        if (!nameFr) nameFr = (c.name && c.name.common) || 'Unknown';
        
        let nameEn = (c.name && c.name.common) || '';
        if (nameFr === 'Unknown') continue;
        
        const root = c.idd.root;
        const suffixes = c.idd.suffixes || [];
        const dialCode = root + (suffixes.length === 1 ? suffixes[0] : '');

        const flag = c.flag || '';
        
        formatted.push({ code, nameFr, nameEn, dialCode, flag });
    }

    formatted.sort((a, b) => a.nameFr.localeCompare(b.nameFr, 'fr', { sensitivity: 'base' }));

    const fileContent = `export type CountryData = {
    code: string;
    nameFr: string;
    nameEn: string;
    dialCode: string;
    flag: string;
};

export const countries: CountryData[] = ${JSON.stringify(formatted, null, 4)};\n`;

    const outPath = path.join(__dirname, 'src', 'utils', 'countries.ts');
    fs.mkdirSync(path.dirname(outPath), { recursive: true });
    fs.writeFileSync(outPath, fileContent, 'utf-8');
    console.log('Successfully wrote', formatted.length, 'countries to', outPath);
}

generate();
